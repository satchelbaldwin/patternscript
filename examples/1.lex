["target/debug/patternscript", "examples/1.pattern"]
Keyword(Bullet)
Id("mid_sized")
Assign
OpenBlock
Id("sprite")
Assign
String("gameasset")
Semicolon
Id("hitbox")
Assign
OpenParen
Number("4")
Comma
Number("4")
CloseParen
Semicolon
Id("color")
Assign
OpenParen
Number("255")
Comma
Number("255")
Comma
Number("0")
CloseParen
Semicolon
CloseBlock
Keyword(Path)
Id("downward_s_curve")
OpenParen
Id("t")
Comma
Id("speed")
Comma
Id("offset")
CloseParen
Assign
OpenBlock
Id("x")
Assign
OpenParen
Number("50")
Operator(Mul)
Id("sin")
OpenParen
Id("t")
CloseParen
CloseParen
Operator(Add)
Id("offset")
Semicolon
Id("y")
Assign
Id("t")
Operator(Mul)
Id("speed")
Semicolon
CloseBlock
Keyword(Pattern)
Id("phase1")
Assign
OpenBlock
Id("iteration_type")
Assign
Id("time")
Semicolon
Comment
Id("length")
Assign
Number("6.0")
Keyword(Seconds)
Semicolon
Comment
Id("actions")
Assign
OpenBlock
Keyword(Let)
Operator(Or)
Id("rigin")
Assign
Id("entity_position")
Operator(Add)
OpenParen
Number("0")
Comma
Number("20")
CloseParen
Semicolon
Keyword(For)
OpenParen
Id("n")
Assign
Number("0")
RangeSeparator
Number("5")
CloseParen
OpenBlock
Keyword(Let)
Id("angle")
Assign
Id("towards_player")
Semicolon
Keyword(Spawn)
OpenBlock
Id("type")
Assign
Id("mid_sized")
Comma
Id("position")
Assign
Operator(Or)
Id("rigin")
Comma
Id("direction")
Assign
Id("angle")
Operator(Add)
Id("n")
Operator(Mul)
Number("20")
Comma
Comment
Id("speed")
Assign
Number("200")
Comma
CloseBlock
Keyword(Wait)
Number("4")
Keyword(Frames)
Semicolon
CloseBlock
Keyword(Wait)
Number("0.5")
Keyword(Seconds)
Semicolon
Keyword(For)
OpenParen
Id("i")
Assign
Number("0")
RangeSeparator
Number("3")
Comma
Id("j")
Assign
Number("0")
RangeSeparator
Number("3")
CloseParen
Condition(Unless)
OpenParen
Id("i")
Operator(Test)
Number("1")
Operator(And)
Id("j")
Operator(Test)
Number("1")
CloseParen
OpenBlock
Keyword(Spawn)
OpenBlock
Id("type")
Assign
Id("mid_sized")
Comma
Id("position_fn")
Assign
Id("downward_s_curve")
OpenParen
Id("t")
Comma
Number("200")
Comma
OpenParen
Id("entity_position")
Operator(Add)
OpenParen
OpenParen
Id("i")
Operator(Sub)
Number("1")
CloseParen
Operator(Mul)
Number("80")
Comma
OpenParen
Id("j")
Operator(Sub)
Number("1")
CloseParen
Operator(Mul)
Number("80")
CloseParen
CloseParen
CloseParen
Comma
CloseBlock
CloseBlock
Keyword(Wait)
Number("0.5")
Keyword(Seconds)
Semicolon
CloseBlock
CloseBlock
Keyword(Pattern)
Semicolon
Keyword(Pattern)
Semicolon
EOF
