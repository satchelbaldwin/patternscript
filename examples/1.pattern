bullet mid_sized = {
     sprite = "gameasset";
     hitbox = (4, 4);
     shape = "rectangle";
     color = (255, 255, 0);
     lifetime = 120;
}

path downward_s_curve(t, speed, offset) = {
     x = (50 * sin(t / 200)) + x(offset);
     y = (t / 200) * speed;
}

pattern phase1 = {
     iteration_type = time;          // time instead of a cycle count
     length = 50.0 seconds;           // time in seconds or frames
     actions = {
          origin = entity_position + (0, 20);

          for (n = 0...100) {
               angle = towards_player;
               spawn {
                    type = mid_sized;
                    position = origin;
                    rotation = angle + n;        // should handle precedence correctly here
                    speed = 500;
                    lifetime = 120;
               }
               spawn {
                    type = mid_sized;
                    position = origin;
                    rotation = angle - n;        // should handle precedence correctly here
                    speed = 500;
               }
               wait 1 frames;
          }
          wait 5 frames;

          //for (i = 0...3, j = 0...3) unless (i == 1 and j == 1) {
          //     spawn {
          //          type = mid_sized;
          //          lifetime = 800;
          //          position_fn = downward_s_curve(t, 200, (entity_position + ((i - 1) * 80, (j - 1) * 80))); 
          //     }
          //}
          //wait 5 frames;
     }
}

