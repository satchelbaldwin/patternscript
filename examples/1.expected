Head(HeadData { 
    definitions: {
        "mid_sized": Bullet(BulletData { 
            definitions: {
                "color": Vector([Int(255), Int(255), Int(0)]), 
                "hitbox": Vector([Int(4), Int(4)]), 
                "sprite": String("gameasset")
            } 
        }), 
        "downward_s_curve": Path(PathData { 
            definitions: {
                "x": Expr(Binary(Add, 
                    Expr(Binary(Mul, 
                        Int(50), 
                        Expr(Unary(FunctionCall("sin"), Variable("t")))
                    )), 
                    Variable("offset")
                )), 
                "y": Expr(Binary(Mul, 
                    Variable("t"), 
                    Variable("speed")
                ))
            } 
        }), 
        "p2": Pattern(PatternData { 
            block: Block { 
                definitions: {
                    "actions": Block(Block { 
                        definitions: {}, 
                        statements: [Spawn(SpawnData { definitions: {} })]
                    })
                }, 
                statements: [] 
            } 
        }), 
        "phase1": Pattern(PatternData { 
            block: Block { 
                definitions: {
                    "length": Duration(Time(Float(6.0))), 
                    "iteration_type": Variable("time"), 
                    "actions": Block(Block { 
                        definitions: {
                            "origin": Expr(Binary(Add, 
                                Variable("entity_position"), 
                                Vector([Int(0), Int(20)]))
                            )
                        }, 
                        statements: [
                            For(ForData { 
                                initial_definitions: {
                                    "n": Range(0, 5)
                                }, 
                                condition: None, 
                                body: Block { 
                                    definitions: {
                                        "angle": Variable("towards_player")
                                    }, 
                                    statements: [
                                        Spawn(SpawnData { 
                                            definitions: {
                                                "position": Variable("origin"), 
                                                "speed": Int(200), 
                                                "type": Variable("mid_sized"), 
                                                "direction": Expr(Binary(Add, 
                                                    Variable("angle"), 
                                                    Expr(Binary(Mul, 
                                                        Variable("n"), 
                                                        Int(20))
                                                    )
                                                ))
                                            } 
                                        }), 
                                        Wait(Frames(Int(4)))
                                    ] 
                                } 
                            }), 
                            Wait(Time(Float(0.5))), 
                            For(ForData { 
                                initial_definitions: {
                                    "j": Range(0, 3), 
                                    "i": Range(0, 3)
                                }, 
                                condition: Unless(
                                    Expr(Binary(And, 
                                        Expr(Binary(Test, 
                                            Variable("i"), 
                                            Int(1)
                                        )), 
                                        Expr(Binary(Test, 
                                            Variable("j"), 
                                            Int(1)
                                        ))
                                    ))
                                ), 
                                body: Block { 
                                    definitions: {
                                        "x": Int(1)
                                    }, 
                                    statements: [
                                        Spawn(SpawnData { 
                                            definitions: {
                                                "type": Variable("mid_sized"), 
                                                "position_fn": Expr(Unary(
                                                    FunctionCall("downward_s_curve"), 
                                                    Vector([
                                                        Variable("t"), 
                                                        Int(200), 
                                                        Expr(Binary(Add, 
                                                            Variable("entity_position"), 
                                                            Vector([
                                                                Expr(Binary(Mul, 
                                                                    Expr(Binary(Sub, 
                                                                        Variable("i"), 
                                                                        Int(1)
                                                                    )), 
                                                                    Int(80)
                                                                )), 
                                                                Expr(Binary(Mul, 
                                                                    Expr(Binary(Sub, 
                                                                        Variable("j"), 
                                                                        Int(1))), 
                                                                    Int(80)
                                                                ))
                                                            ])
                                                        ))
                                                    ])
                                                ))
                                            } 
                                        })
                                    ] 
                                } 
                            }), 
                            Wait(Time(Float(0.5)))
                        ] 
                    }
                )}, 
                statements: [] 
            } 
        })
    } 
})